src/middlewared/middlewared/service/config_service.py

```diff
@@ -119,12 +119,14 @@ async def update(self, app, message_id, data):
 
     @private
     async def _get_or_insert(self, datastore, options):
-        try:
-            return await self.middleware.call('datastore.config', datastore, options)
-        except IndexError:
+        rows = await self.middleware.call('datastore.query', datastore, [], options)
+        if not rows:
             async with get_or_insert_lock:
-                try:
-                    return await self.middleware.call('datastore.config', datastore, options)
-                except IndexError:
+                # We do this again here to avoid TOCTOU as we don't want multiple calls inserting records
+                # and we ending up with duplicates again
+                rows = await self.middleware.call('datastore.query', datastore, [], options)
```

Comment:
I'd also like to see a comment about former behavior and why simple:
```
try:
except IndexError:
```
doesn't work and can lead to corruption.